### PR TITLE
Suggest using anonymous lifetime in `impl Trait` return

### LINKED ISSUE #48467: impl trait error message around `'static` is probably confusing
The error message for [this example](https://play.rust-lang.org/?gist=6d2e6982214b36054a9716f80335f0c0&version=nightly) seems potentially confusing:

```rust
#![feature(dyn_trait)]
#![feature(conservative_impl_trait)]
#![feature(underscore_lifetimes)]

fn a<T>(items: &[T]) -> Box<impl Iterator> {
    Box::new(items.iter()) //~ ERROR cannot infer an appropriate lifetime
}

fn main() { }
```

I get: 

```
error[E0621]: explicit lifetime required in the type of `items`
 --> src/main.rs:5:29
  |
5 | fn a<T>(items: &[T]) -> Box<impl Iterator> {
  |         -----               ^^^^^^^^^^^^^ lifetime `'static` required
  |         |
  |         consider changing the type of `items` to `&'static [T]`
```

What's going wrong here is that `impl Iterator` can only capture lifetimes that it names, and this one doesn't name any lifetimes. It can be fixed by changing to `impl Iterator + '_` -- this is probably what we want to suggest? I'm not sure how best to phrase this, suggestions welcome.

cc @cramertj @estebank

### PR DESCRIPTION
**Pull request by estebank**
Fix #48467.

r? @nikomatsakis

### CODE REVIEW

**Comment by estebank** [comment]
This is a blatant hack, but wanted to have a PR out so we can talk about the desired output.

**Comment by pnkfelix** [comment]
Hmm that output is interesting.

I wonder, as a matter of coding style and/or readability, if people would prefer to see the following suggested:
```rust
 fn elided<'a>(x: &'a i32) -> impl Copy + 'a { x }
```

(but I don't know, I think I often err on the side of explicitness when people would prefer elision...)

**Comment by pnkfelix** [comment]
Oh; I guess that test case, in its full form, even includes an example where we *do* suggest that, if you start with `fn elided<'a>(x: &'a i32) -> impl Copy { x }`.

So its really a matter of style, and I suspect the prevailing style is towards what @estebank is proposing here

**Comment by estebank** [comment]
We can expand this code to also suggest

```
help: you can add an explicit named lifetime `'a` to the type of `x` and the return type
   |
LL | fn elided<a>(x: &'a i32) -> impl Copy + 'a { x }
   |          ^^^     ^^                   ^^^^
```

but the code itself to make that suggestion might be quite brittle.

**Comment by Centril** [comment]
I think the form suggested here is quite reasonable and I suspect doing the more elaborate suggestion is more complex impl wise...

To reduce Niko's backlog, r? @pnkfelix

**Comment by Centril** [comment]
Ping from triage, @pnkfelix -- shall we go with @estebank's idea for now?

**Comment by estebank** [comment]
I had a talk with Niko and he clarified what a better approach would be.

I looked around and arrived to the current solution, with which I'm relatively happy with (even though I think we could improve the `static_impl_trait` output in a separate PR). I feel that this PR is now reasonable/reviewable/mergeable.

File: src/test/ui/impl-trait/must_outlive_least_region_or_bound.stderr [hunk 1]
```diff
@@ -1,10 +1,20 @@
...
+   |                       ---------   ^ ...but this borrow...
+   |                       |
+   |                       this return type evaluates to the `'static` lifetime...
+   |
+note: ...can't outlive the anonymous lifetime #1 defined on the function body at 3:1
+  --> $DIR/must_outlive_least_region_or_bound.rs:3:1
+   |
+LL | fn elided(x: &i32) -> impl Copy { x }
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+help: you can add a constraint to the return type to make it last less than `'static` and match the anonymous lifetime #1 defined on the function body at 3:1
```
Comment (Centril) [inline src/test/ui/impl-trait/must_outlive_least_region_or_bound.stderr]: Btw... "constraint" here is odd; we typically use "bound" for things in `A + B+ ...`.

Comment (estebank) [reply src/test/ui/impl-trait/must_outlive_least_region_or_bound.stderr]: We've been using that wording for 6 months, but I agree we could improve on the wording on these errors.

File: src/librustc/infer/error_reporting/nice_region_error/named_anon_conflict.rs [hunk 2]
```diff
@@ -84,6 +86,13 @@ impl<'a, 'gcx, 'tcx> NiceRegionError<'a, 'gcx, 'tcx> {
             {
                 return None;
             }
+            if let FunctionRetTy::Return(ty) = &fndecl.output {
+                if let (TyKind::Def(_, _), ty::ReStatic) = (&ty.node, sub) {
+                    // This is an impl Trait return that evaluates de need of 'static.
```
Comment (pnkfelix) [inline src/librustc/infer/error_reporting/nice_region_error/named_anon_conflict.rs]: "evaluates _de_ need of 'static" ?

What is this?

Comment (estebank) [reply src/librustc/infer/error_reporting/nice_region_error/named_anon_conflict.rs]: A partial rewording of the prior comment 😬

I believe something along the lines of "that evaluates to a need of static".

**Pull request merged**